classes/data-stores/ActionScheduler_DBStore.php

```diff
@@ -0,0 +1,625 @@
+			if ( empty( $group_id ) ) {
+				throw new InvalidArgumentException( sprintf( __( 'The group "%s" does not exist.', 'action-scheduler' ), $group ) );
+			}
+
+			$where   .= ' AND group_id = %d';
+			$params[] = $group_id;
+		}
+
+		$order    = "ORDER BY attempts ASC, scheduled_date_gmt ASC, action_id ASC LIMIT %d";
+		$params[] = $limit;
+
+		$sql = $wpdb->prepare( "{$update} {$where} {$order}", $params );
+
+		$rows_affected = $wpdb->query( $sql );
+		if ( $rows_affected === false ) {
+			throw new \RuntimeException( __( 'Unable to claim actions. Database error.', 'action-scheduler' ) );
+		}
+
+		return (int) $rows_affected;
+	}
+
+	/**
+	 * @return int
+	 */
+	public function get_claim_count() {
+		global $wpdb;
+
+		$sql = "SELECT COUNT(DISTINCT claim_id) FROM {$wpdb->actionscheduler_actions} WHERE claim_id != 0 AND status IN ( %s, %s)";
+		$sql = $wpdb->prepare( $sql, [ self::STATUS_PENDING, self::STATUS_RUNNING ] );
+
+		return (int) $wpdb->get_var( $sql );
+	}
+
+	/**
+	 * Return an action's claim ID, as stored in the claim_id column
+	 *
+	 * @param string $action_id
+	 * @return mixed
+	 */
+	public function get_claim_id( $action_id ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+
+		$sql = "SELECT claim_id FROM {$wpdb->actionscheduler_actions} WHERE action_id=%d";
+		$sql = $wpdb->prepare( $sql, $action_id );
+
+		return (int) $wpdb->get_var( $sql );
+	}
+
+	/**
+	 * @param string $claim_id
+	 *
+	 * @return int[]
+	 */
+	public function find_actions_by_claim_id( $claim_id ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+
+		$sql = "SELECT action_id FROM {$wpdb->actionscheduler_actions} WHERE claim_id=%d";
+		$sql = $wpdb->prepare( $sql, $claim_id );
+
+		$action_ids = $wpdb->get_col( $sql );
+
+		return array_map( 'intval', $action_ids );
+	}
+
+	public function release_claim( ActionScheduler_ActionClaim $claim ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+		$wpdb->update( $wpdb->actionscheduler_actions, [ 'claim_id' => 0 ], [ 'claim_id' => $claim->get_id() ], [ '%d' ], [ '%d' ] );
+		$wpdb->delete( $wpdb->actionscheduler_claims, [ 'claim_id' => $claim->get_id() ], [ '%d' ] );
+	}
+
+	/**
+	 * @param string $action_id
+	 *
+	 * @return void
+	 */
+	public function unclaim_action( $action_id ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+		$wpdb->update(
+			$wpdb->actionscheduler_actions,
+			[ 'claim_id' => 0 ],
+			[ 'action_id' => $action_id ],
+			[ '%s' ],
+			[ '%d' ]
+		);
+	}
+
+	public function mark_failure( $action_id ) {
+		/** @var \wpdb $wpdb */
+		global $wpdb;
+		$updated = $wpdb->update(
+			$wpdb->actionscheduler_actions,
+			[ 'status' => self::STATUS_FAILED ],
+			[ 'action_id' => $action_id ],
+			[ '%s' ],
+			[ '%d' ]
+		);
+		if ( empty( $updated ) ) {
+			throw new \InvalidArgumentException( sprintf( __( 'Unidentified action %s', 'action-scheduler' ), $action_id ) );
+		}
+	}
+
+	/**
+	 * @param string $action_id
+	 *
+	 * @return void
+	 */
+	public function log_execution( $action_id ) {
```

Comment:
The reason for the query instead of the update function is `attempts = attempts+1`. If you retrieve the record, increment & save it's possible to overwrite an update that occurred while in process.